### PR TITLE
Add fine grained yolov3 loss

### PR DESCRIPTION
**Pull request by heavengate**
**Add splited yolov3 loss**
- Add Gt2Target process in post_map.py
- Add yolo_loss in ppdet/modeling/losses/yolo_loss.py
- Train YOLOv3 with fine grainted loss by
```
python tools/train.py -c configs/yolov3_darknet_voc.yml -o use_fine_grained_loss=true
```

### CODE REVIEW

File: ppdet/modeling/anchor_heads/yolo_head.py [hunk 1]
```diff
@@ -269,7 +278,147 @@ def get_loss(self, input, gt_box, gt_label, gt_score):
             losses.append(fluid.layers.reduce_mean(loss))
             downsample //= 2
 
-        return sum(losses)
+        return {'loss': sum(losses)}
+
+    def get_splited_loss(self, outputs, targets, gt_box):
```
Comment (qingqing01) [inline ppdet/modeling/anchor_heads/yolo_head.py]: maybe create `losses` folder in `ppdet/modeling`,  put this loss in that folder.

File: ppdet/modeling/anchor_heads/yolo_head.py [hunk 2]
```diff
@@ -269,7 +278,147 @@ def get_loss(self, input, gt_box, gt_label, gt_score):
...
+            loss_h = fluid.layers.reduce_sum(loss_h, dim=[1, 2, 3])
+
+            # A prediction bbox overlap any gt_bbox over ignore_thresh, 
+            # objectness loss will be ignored, process as follows:
+
+            # 1. get pred bbox, which is same with YOLOv3 infer mode, use yolo_box here
+            # NOTE: img_size is set as 1.0 to get noramlized pred bbox
+            bbox, _ = fluid.layers.yolo_box(
+                x=output,
+                img_size=fluid.layers.ones(shape=[self.batch_size, 2], dtype="int32"),
```
Comment (qingqing01) [inline ppdet/modeling/anchor_heads/yolo_head.py]: 如果训练中最后一个iter不足batch_size个，这里会不会出错

File: ppdet/modeling/anchor_heads/yolo_head.py [hunk 3]
```diff
@@ -243,13 +247,18 @@ def get_loss(self, input, gt_box, gt_label, gt_score):
...
+            targets ([Variables]): List of Variables, the targets for yolo
+                                   loss calculatation.
 
         Returns:
             loss (Variable): The loss Variable of YOLOv3 network.
 
         """
         outputs = self._get_outputs(input, is_train=True)
 
+        if self._use_splited_loss:
```
Comment (qingqing01) [inline ppdet/modeling/anchor_heads/yolo_head.py]: use_splited_loss ->  use_fine_grained _loss

File: ppdet/modeling/anchor_heads/yolo_head.py (hunk 1 above)
Comment (heavengate) [reply ppdet/modeling/anchor_heads/yolo_head.py]: Done, Thanks!

File: ppdet/modeling/anchor_heads/yolo_head.py (hunk 3 above)
Comment (heavengate) [reply ppdet/modeling/anchor_heads/yolo_head.py]: Done, Thanks!

File: ppdet/modeling/anchor_heads/yolo_head.py (hunk 2 above)
Comment (heavengate) [reply ppdet/modeling/anchor_heads/yolo_head.py]: 是的，目前好像没有动态把batch size维split成1或者对batch size维for循环的方式，train的时候一般都会drop_last，通过这个方式规避下？

File: ppdet/data/data_feed.py [hunk 4]
```diff
@@ -34,7 +34,7 @@
     ArrangeTestSSD, ArrangeYOLO, ArrangeEvalYOLO, ArrangeTestYOLO)
 
 __all__ = [
-    'PadBatch', 'MultiScale', 'RandomShape', 'PadMSTest', 'DataSet',
+    'PadBatch', 'MultiScale', 'RandomShape', 'Gt2Target', 'PadMSTest', 'DataSet',
```
Comment (jerrywgz) [inline ppdet/data/data_feed.py]: Gt2YoloTarget? This operator is also needed In AnchorFree models

Comment (heavengate) [reply ppdet/data/data_feed.py]: Done, Thanks!

**Comment by wangchaochaohu** [comment]
Great

**Comment by littletomatodonkey** [comment]
LGTM~

File: tools/train.py [hunk 5]
```diff
@@ -118,6 +118,12 @@ def main():
             model = create(main_arch)
             train_loader, feed_vars = create_feed(train_feed)
 
+            # YOLOv3 splited loss code need batch_size for calculation set
+            # yolo_head.yolo_loss.batch_size by batch_size in YoloTrainFeed here
+            if main_arch == "YOLOv3" and getattr(cfg, 'use_fine_grained_loss',
+                                                 False):
+                model.yolo_head.yolo_loss.batch_size = train_feed.batch_size
```
Comment (qingqing01) [inline tools/train.py]: 这种参数可以配置在yml文件里，而不是train.py里加规则

File: ppdet/modeling/anchor_heads/yolo_head.py [hunk 6]
```diff
@@ -34,38 +35,38 @@ class YOLOv3Head(object):
...
 
     def __init__(self,
                  norm_decay=0.,
                  num_classes=80,
-                 ignore_thresh=0.7,
-                 label_smooth=True,
                  anchors=[[10, 13], [16, 30], [33, 23], [30, 61], [62, 45],
                           [59, 119], [116, 90], [156, 198], [373, 326]],
                  anchor_masks=[[6, 7, 8], [3, 4, 5], [0, 1, 2]],
+                 yolo_loss="YOLOv3Loss",
```
Comment (qingqing01) [inline ppdet/modeling/anchor_heads/yolo_head.py]: 这里是类似 nms, backone等模块，如下，可以配置MultiClassSoftNMS，也可以配置MultiClassNMS.

```
CascadeBBoxHead:
  head: CascadeTwoFCHead
  nms: MultiClassSoftNMS

CascadeTwoFCHead:
  mlp_dim: 1024

MultiClassSoftNMS:
  score_threshold: 0.01
  keep_top_k: 300
  softnms_sigma: 0.5
```

我觉得是， yolo_loss可以配置 YOLOv3Loss,  YOLOv3改进的loss。
默认的yolov3-loss里依据`use_fine_grained_loss`选择用cpp-loss，还是py-loss。

File: ppdet/modeling/losses/yolo_loss.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+                 batch_size=8,
+                 ignore_thresh=0.7,
+                 label_smooth=True,
+                 use_fine_grained_loss=False):
+        self.batch_size = batch_size
+        self._ignore_thresh = ignore_thresh
+        self._label_smooth = label_smooth
+        self._use_fine_grained_loss = use_fine_grained_loss
+
+    def get_loss(self, outputs, gt_box, gt_label, gt_score, targets, anchors,
```
Comment (qingqing01) [inline ppdet/modeling/losses/yolo_loss.py]: `def __calll__` 而不是 `get_loss`

Comment (qingqing01) [inline ppdet/modeling/losses/yolo_loss.py]: `def __calll__` 而不是 `get_loss`

File: ppdet/modeling/anchor_heads/yolo_head.py (hunk 6 above)
Comment (qingqing01) [inline ppdet/modeling/anchor_heads/yolo_head.py]: 这里是类似 nms, backone等模块，如下，可以配置MultiClassSoftNMS，也可以配置MultiClassNMS.

```
CascadeBBoxHead:
  head: CascadeTwoFCHead
  nms: MultiClassSoftNMS

CascadeTwoFCHead:
  mlp_dim: 1024

MultiClassSoftNMS:
  score_threshold: 0.01
  keep_top_k: 300
  softnms_sigma: 0.5
```

觉得是， yolo_loss可以配置 YOLOv3Loss,  YOLOv3改进的loss。
默认的yolov3-loss里依据`use_fine_grained_loss`选择用cpp-loss，还是py-loss。

Comment (qingqing01) [inline ppdet/modeling/anchor_heads/yolo_head.py]: 这里是类似 nms, backone等模块，如下，可以配置MultiClassSoftNMS，也可以配置MultiClassNMS.

```
CascadeBBoxHead:
  head: CascadeTwoFCHead
  nms: MultiClassSoftNMS

CascadeTwoFCHead:
  mlp_dim: 1024

MultiClassSoftNMS:
  score_threshold: 0.01
  keep_top_k: 300
  softnms_sigma: 0.5
```

觉得是， yolo_loss可以配置 YOLOv3Loss,  YOLOv3改进的loss。
默认的yolov3-loss里依据`use_fine_grained_loss`选择用cpp-loss，还是py-loss。

Comment (heavengate) [reply ppdet/modeling/anchor_heads/yolo_head.py]: Done, Thanks
